zap/src/irgen/gen.rs

```diff
@@ -1,517 +1,603 @@
+			stmts.push(Stmt::NumFor {
+				var: "_",
+				from: 0.0.into(),
+				to: buffer_readu16(),
+			});
 
-				self.local("val");
-				self.des(val, &"val".into());
+			local(&mut stmts, "k", None);
+			local(&mut stmts, "v", None);
 
-				self.assign(into.clone().expr_index("key".into()), "val".into());
+			stmts.extend(gen_des(key, "k".into(), gen_checks));
+			stmts.extend(gen_des(val, "v".into(), gen_checks));
 
-				self.block_end();
+			assign(&mut stmts, to.clone().eindex("k".into()), "v".into());
 
-				self.block_end();
-			}
+			stmts.push(Stmt::End);
+		}
 
-			Ty::Struct { fields } => {
-				self.assign(into.clone(), Expr::EmptyObj);
+		Ty::Struct { fields } => {
+			assign(&mut stmts, to.clone(), Expr::EmptyTab);
 
-				for (name, ty) in fields {
-					self.des(ty, &into.clone().name_index(name.clone()));
-				}
+			for (name, ty) in fields {
+				stmts.extend(gen_des(ty, to.clone().nindex(name), gen_checks));
 			}
+		}
 
-			Ty::Enum { variants } => {
-				let num_ty = NumTy::from_f64(0.0, variants.len() as f64 - 1.0);
-
-				self.block_start();
+		Ty::Enum { variants } => {
+			if !variants.is_empty() {
+				let numty = NumTy::from_f64(0.0, variants.len() as f64 - 1.0);
 
-				self.local("val");
-				self.read_num("val".into(), num_ty);
+				assign(&mut stmts, to.clone(), buffer_readnumty(numty));
 
-				for (i, variant) in variants.iter().enumerate() {
+				for (i, name) in variants.iter().enumerate() {
 					if i == 0 {
-						self.if_(Expr::from("val").eq(i.into()));
+						stmts.push(Stmt::If(Expr::from(to.clone()).eq((i as f64).into())));
 					} else {
-						self.else_if(Expr::from("val").eq(i.into()));
+						stmts.push(Stmt::ElseIf(Expr::from(to.clone()).eq((i as f64).into())));
 					}
 
-					self.assign(into.clone(), variant.clone().into());
+					assign(&mut stmts, to.clone(), name.clone().into());
 				}
 
-				self.else_();
-
-				self.throw("Invalid enum variant!".into());
-
-				self.block_end();
-
-				self.block_end();
+				stmts.push(Stmt::Else);
+				stmts.push(Stmt::Error("invalid enum variant!".to_string()));
+				stmts.push(Stmt::End);
 			}
+		}
 
-			Ty::Instance(_) => {
-				self.read_inst(into.clone());
-				self.assert(Expr::from(into.clone()), "Instance could not be found!".into());
-			}
-
-			Ty::Vector3 => {
-				self.local("x");
-				self.local("y");
-				self.local("z");
-
-				self.read_num("x".into(), NumTy::F32);
-				self.read_num("y".into(), NumTy::F32);
-				self.read_num("z".into(), NumTy::F32);
+		Ty::Instance(class) => {
+			assign(
+				&mut stmts,
+				to.clone(),
+				Var::from("incoming_inst").eindex(buffer_readu16()).into(),
+			);
 
-				self.assign(
-					into.clone(),
-					Expr::Vector3(Box::new("x".into()), Box::new("y".into()), Box::new("z".into())),
+			// Assert that the instance is not nil even if we don't want checks
+			// because roblox cannot ensure the instance's existance at the destination
+			assert(&mut stmts, Expr::from(to.clone()).neq(Expr::Nil), None);
```

Comment:
Is this the right way to go? As an error will... error when we might want to handle gracefully?